server/src/main/java/org/opensearch/index/engine/AbstractNRTReplicationEngine.java

```diff
@@ -0,0 +1,370 @@
+            this.readerManager.addListener(completionStatsCache);
+            for (ReferenceManager.RefreshListener listener : engineConfig.getExternalRefreshListener()) {
+                this.readerManager.addListener(listener);
+            }
+            for (ReferenceManager.RefreshListener listener : engineConfig.getInternalRefreshListener()) {
+                this.readerManager.addListener(listener);
+            }
+            final Map<String, String> userData = store.readLastCommittedSegmentsInfo().getUserData();
+            final String translogUUID = Objects.requireNonNull(userData.get(Translog.TRANSLOG_UUID_KEY));
+            translogManagerRef = createTranslogManager(translogUUID, translogManager);
+            translogManager.set(translogManagerRef);
+        } catch (IOException e) {
+            IOUtils.closeWhileHandlingException(store::decRef, readerManager);
+            closeIfCloseable(IOUtils::closeWhileHandlingException, translogManagerRef);
+            throw new EngineCreationFailureException(shardId, "failed to create engine", e);
+        }
+    }
+
+    @Override
+    public IndexResult index(Index index) throws IOException {
+        ensureOpen();
+        IndexResult indexResult = new IndexResult(index.version(), index.primaryTerm(), index.seqNo(), false);
+        final Translog.Location location = translogManager().add(new Translog.Index(index, indexResult));
+        indexResult.setTranslogLocation(location);
+        indexResult.setTook(System.nanoTime() - index.startTime());
+        indexResult.freeze();
+        getLocalCheckpointTracker().advanceMaxSeqNo(index.seqNo());
+        return indexResult;
+    }
+
+    @Override
+    public DeleteResult delete(Delete delete) throws IOException {
+        ensureOpen();
+        DeleteResult deleteResult = new DeleteResult(delete.version(), delete.primaryTerm(), delete.seqNo(), true);
+        final Translog.Location location = translogManager().add(new Translog.Delete(delete, deleteResult));
+        deleteResult.setTranslogLocation(location);
+        deleteResult.setTook(System.nanoTime() - delete.startTime());
+        deleteResult.freeze();
+        getLocalCheckpointTracker().advanceMaxSeqNo(delete.seqNo());
+        return deleteResult;
+    }
+
+    @Override
+    public NoOpResult noOp(NoOp noOp) throws IOException {
+        ensureOpen();
+        NoOpResult noOpResult = new NoOpResult(noOp.primaryTerm(), noOp.seqNo());
+        final Translog.Location location = translogManager().add(new Translog.NoOp(noOp.seqNo(), noOp.primaryTerm(), noOp.reason()));
+        noOpResult.setTranslogLocation(location);
+        noOpResult.setTook(System.nanoTime() - noOp.startTime());
+        noOpResult.freeze();
+        getLocalCheckpointTracker().advanceMaxSeqNo(noOp.seqNo());
+        return noOpResult;
+    }
+
+    protected abstract TranslogManager createTranslogManager(String translogUUID, SetOnce<TranslogManager> translogManager)
```

Comment:
Can we delegate the creation of the TranslogManager to a factory and inject that rather than these 3 NRT engine types?

Reply:
So, had to create an abstract engine so that the seg rep can still work for the newer engine. And since there are some other engine methods that had to be overridden, there came the need to create a newer engine altogether. 
Around the creation of TranslogManager, this should be doable.